src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -1704,6 +1721,182 @@ case class MatrixFilterEntries(child: MatrixIR, pred: IR) extends MatrixIR {
   }
 }
 
+case class MatrixAnnotateRowsTable(
+  child: MatrixIR,
+  table: TableIR,
+  root: String,
+  key: Option[IndexedSeq[IR]]) extends MatrixIR {
+
+  def children: IndexedSeq[BaseIR] = FastIndexedSeq(child, table) ++ key.getOrElse(FastIndexedSeq.empty[IR])
+
+  override def columnCount: Option[Call] = child.columnCount
+
+  override def partitionCounts: Option[IndexedSeq[Long]] = child.partitionCounts
+
+  val typ: MatrixType = child.typ.copy(rvRowType = child.typ.rvRowType ++ TStruct(root -> table.typ.valueType))
+
+  def copy(newChildren: IndexedSeq[BaseIR]): BaseIR = {
+    MatrixAnnotateRowsTable(
+      newChildren(0).asInstanceOf[MatrixIR],
+      newChildren(1).asInstanceOf[TableIR],
+      root,
+      key.map { keyIRs =>
+        assert(newChildren.length == keyIRs.length + 2)
+        newChildren.drop(2).map(_.asInstanceOf[IR])
+      }
+    )
+  }
+
+  override def execute(hc: HailContext): MatrixValue = {
+    val prev = child.execute(hc)
+    val tv = table.execute(hc)
+    key match {
+      // annotateRowsIntervals
+      case None if child.typ.rowPartitionKey.length == 1 &&
```

Comment:
This whole block was moved over from MatrixTable.scala as-is for parsimony, and I don't really want to do a full review-and-change cycle for code I didn't actually modify. I'm sorry I wasn't clear about that at the beginning!

Reply:
👍